### DISCUSSION TITLE
Is anyone running Postal v2 as a production server?

### DISCUSSION THREAD

**Discussion by pjv** [postalserver/postal#1612] [Help with using Postal]
I have a working v1 Postal install on an under-powered VPS that can send ~ 200 emails/minute. I’m happy that Postal has been getting a lot of upgrade love lately and looking forward to moving on to v2. To that end, I’ve provisioned a new VPS and installed Postal v2. I’ve successfully tested migrating a backup of my v1 database to the v2 server. 

Although the [docs](https://docs.postalserver.io/install/upgrade-to-v2) mention rolling back the upgrade if v2 isn’t working out, since the upgrade involves a db schema update, that seems like a sketchy procedure so I’m considering the upgrade from v1 to v2 to be a one-way thing.

My postal server is handling transactional email for several busy ecommerce sites and I need to keep down-time low, not lose any emails (which are often payment receipts to customers), and can’t afford to move to v2 and find it’s unable to keep up.

With that context, and wanting some confidence before I pull the upgrade trigger, my question is whether anyone is successfully running a production Postal v2 deployment to deliver a high volume of email. 

Followup question for anyone who answers affirmative to that is whether there were any adjustments beyond what is covered in the standard documentation to get to a working config.

**Comment by pjv** [comment]
I changed the title of the thread from being about throughput to just being about whether v2 is stable enough to use in production.

**Reply by willpower232** [reply]
I'd presume krystal are running it in production, we haven't had an opportunity yet.

I think a major consideration is in the number of workers, you'll probably want to scale them to a higher number when you first run it as the previous config setting is not involved any more I think. You can probably prove the process count without moving production over to be more sure that you have a similar number of processes.

**Comment by ndTEC** [comment]
We migrate to v2 in production. The most important problem we had was with error dh key to small because of openssl (see this: #1641  ) version from worker containers, other than that it works fine, we scale to 4 workers.
We ended up not using the last settings from openssl.cnf

**Reply by pjv** [reply]
Thanks for posting. Can I ask what kind of volume (messages per minute) you are passing through your v2 postal server?

**Comment by ndTEC** [comment]
Last days we reached 180/minute.

**Comment by pjv** [comment]
A lot of what I am seeing in these discussions and in the issues are what seem to me to be showstopper issues with version 2 and I'm still afraid to try it.

Three questions for @adamcooke:

1. Are you / krystal running Postal v2 as a production email sending service?
2. If so, are you running the version in the `main` branch of this repository - unpatched with any of the un-merged pull requests and work-arounds that have been posted here (e.g. #521)?
3. In the documentation for upgrading from v1 to v2 you mention failing back to v1 if there are any issues with v2, but one of the steps in upgrading to v2 is a schema migration on the database. Is the migrated database really 100% backward compatible with a standard v1 postal install?

**Reply by pjv** [reply]
bumping this to request @adamcooke to please respond.
next bump i will plead instead of request.
after that comes begging.
please don't make me beg.

**Reply by pjv** [reply]
@adamcooke as a courtesy, would you be kind enough to say that you aren't going to answer these questions if you aren't intending to answer them? I am actually waiting on a reply to this to make practical decisions about how to proceed with my Postal installs.

**Reply by adamcooke** [reply]
You seem to have skipped pleading and begging and gone straight to self-righteous demanding, but anyway.... 

1. Yes
2. We use the version provided in the published container images and we are not using any patches from unmerged pull requests. 
3. I do not believe the database migrations are destructive so the database will still run on v1 even after running the v2 migrations. I would still, however, recommend taking a backup or snapshot of your data before upgrading.

**Reply by pjv** [reply]
honestly truly sorry if it came off as either self-righteous or overly demanding. I'm just trying to use Postal in a production setting where I can't easily experiment with some aspects of it without possibly causing a significant customer service kerfluffel. I felt like I needed some information that - as far as I know - only you could provide. I asked those questions 9 days ago and it seemed like maybe they were slipping through your cracks. I appreciate your making Postal available as open source software and do not believe that you are obligated in any way to answer my or anyone's questions. Thank you for taking the time to answer them.

**Reply by pjv** [reply]
In case anyone else is facing the same issue and has the same question about the backwards compatibility of the database schema migration, the reason that I asked question number 3 above is as follows: 

I already know that v.2 of Postal will run and initially work with my migrated v.1 database because I have gotten that far in my testing.

What I don't know, and can't know without taking the plunge is whether it will work well enough (e.g. high enough throughput, no other showstoppers) for the dozen or so clients I have configured using it for their transactional emails; some of them pretty busy. 

Assuming that problems show up after some time (hours or days), the current database on the running v.2 server will have diverged significantly from the backup of the v.1 database. Many emails will have been sent on v.2 that won't exist in the backup. The queues, held, suppression lists, etc. will look totally different. Downgrading back to v.1 with the updated v.2 database, if it works ok, would be a tolerable disruption. But having to bring v.1 of Postal back up with the backed-up v.1 database from some hours or days earlier could result in many emails (password resets, payment receipts, account creation confirmations, etc) being delivered multiple times and/or not being delivered at all. That would not be a tolerable (to me, anyway) disruption if I can find any way to avoid it.

**Reply by willpower232** [reply]
You can see in github that only a couple of migrations have been added recently so as long as you aren't trying to run v2 on a v1 database, you shouldn't have issues.

https://github.com/postalserver/postal/tree/main/db/migrate
